### ISSUE TITLE
move from refLists to arrays

### ISSUE THREAD

**Issue by lefnire** [HabitRPG/habitica#634]
This isn't fully decided, but we're going to explore this route since we've been hitting some issues with refLists in our derby experiences. Here are some copied email conversations so anyone interested can chime in:

@toebu:

> I tried to fix the problem that the tagging is screwed up when you add a new task or remove an existing one. Somehow, the nexted named scopes from the #each don't really seem to work properly. Around here: https://github.com/toebu/habitrpg/blob/tags-filters/views/app/tasks.html#L166. I'm looping over the tags and inside that loop :task.tags is not set properly when adding a new task: all the existing tasks will then have the newly added :task. Found some people that had the same problem: https://github.com/codeparty/derby/issues/37 but it says it has been fixed. Just wanted to ask if you ever had a similar problem? If not, no problem, I'll look into it more, but I imagine that one would have to fix something in derby.
> 
> I gave the tags bug another shot. I'm now certain it's something in derby. I removed all the elements from the page except for the bare minimum to reproduce the bug. It still happens then. I also noticed that the tag jumps to another task if you re-order the tasks.
> 
> I traced the problem to the patchCtx function in derby: https://github.com/lefnire/derby/blob/master/lib/viewPath.js#L357. I think patchCtx is responsible for updating the $indices when a list changes. This doesn't work as it should however. That's all I've got. I wonder if we should move away from refLists, they seem to be causing quite a few tricky bugs and don't seem to be very stable yet.
> Is there any reason we couldn't just store the tasks in an array? Or rather store the habits, dailys, todos and rewards in separate arrays. I somehow think that this will also help us to solve some sorting bugs and get rid of that cron job that restores the refLists.

@lefnire

> I tried pulling in SLaks/racer and Unroll-Me/derby. Those go hand-in-hand, and are both maintained by SLaks who's doing a ton of great refList work for bug fixes and performance improvements. Issues still there though.
> 
> The only problem I see in moving from refLists to arrays is that refLists affords us the ease to look tasks up by id - "_user.tasks.#{taskId}" - which we can't do with arrays. We can do that in code with `_.findWhere habits, {id: id}` - we just need to make sure we're not doing any such id reference in the dom that would break with an array migration. Additionally, the API looks things tasks up simply by id (since they're currently all stored in user.tasks, but distinguished by :type). The API looks like /user/tasks/:taskId. Basically, we'd just need to go through the codebase and make sure everything would still work just fine if we moved from refLists (object storage format, :id => task) to an array.

**Comment by SLaks** [comment]
The `#each` issue looks a lot like https://github.com/codeparty/derby/issues/215.

I worked around that by making the `#each` non-live, which probably won't work for you.  

I think this is fixed by https://github.com/codeparty/derby/pull/204, but I merged that into Unroll-Me/derby, so if you're using my fork, I guess I'm wrong about that.  
(since I worked around it, I didn't actually check whether it was fixed)

Are you sure that this bug happens in the transition from the DOM to model paths (`model.at(elem)` or `e.at()`)? (as opposed to a path bug in Racer)

I don't really understand Derby's binding code (yet), so I can't provide more specific help there.  
(I am fortunate enough to have had little trouble with that)  
However, I don't think switching away from refLists would help; if it's a problem with view mapping, it would happen for both arrays and refLists.

One ugly workaround might be to bind the full model (or the IDs necessary to reconstruct it) to each relevant DOM element in a `data-` attribute, then use that to get a scoped model directly.

**Comment by SLaks** [comment]
What is the "cron job that restores the refLists"?  
What problem is that trying to solve?

I'm thinking of introducing a `classRef` feature that will automatically maintain refs and perhaps also refLists or queries on all objects in a collection.  (this would simplify my code and allow such refs to all share a single refGetter & event handler)

Would that help you?

**Comment by SLaks** [comment]
On an unrelated note, I also plan to make unused queries (with no active references) go away soon.

**Comment by lefnire** [comment]
Ok cool, I'll play around some more with `e.at()` using Unroll-Me/derby and see if I can get something more definitive. We're already using `data-id` throughout the system as a work-around, so I have no qualms using it as necessary. And that's good ("good") to know that the same problem applies to arrays as refLists, so we should just work on getting this pegged.

**Comment by SLaks** [comment]
I'm not sure whether the same problem would apply to arrays, but I don't see why it wouldn't.

Are you sure that this problem is purely in the Derby half?

**Comment by lefnire** [comment]
The cron thing: we run "cron" on the client once each morning, their completed dailies then become unchecked and their incomplete dailies hurt them. That's what that does.

The restoring of refLists bit applies to that original sort-order ticket. For one reason or another, people would load their account and `user.habitIds`, `user.dailyIds`, etc would just be `[]` - completely wiped. Looked like they'd lost all their data (which still existed in `user.tasks`). So I just go through and say, "if it's in `user.tasks` but not in the id-list, put it back". It's terrible :/ Couldn't help but associate it with refLists, they weren't losing anything else

It was very strange too because I did step-debugging (webstorm) of my own account during this time. Load my account, play around - was fine. Refresh, step-debug through the server to try to find where I might be inadvertently wiping id-lists (user.habitIds, etc). Nothing - but it got lost in the refresh.

**Comment by lefnire** [comment]
> On an unrelated note, I also plan to make unused queries go away soon.

Woohoo! Kick all those strays to the curb

**Comment by SLaks** [comment]
Try adding a `mutate` and perhaps `beforeTxn` model event handler to log all model changes and see if you can pinpoint where it's getting reset.

**Comment by SLaks** [comment]
I finished my refList perf work.

It's a major change to refList resolution code, but it works fine on my project (and Racer's unit tests).

Can you try it (pull master) and see if it helps performance, or if it fixes or breaks anything?  
(in particular, my code doesn't actually perform any mutation operations on refLists)

Note that master also contains my [earlier breaking change](https://groups.google.com/forum/?fromgroups=#!topic/derbyjs/NqxC3-HXVwY).

**Comment by SLaks** [comment]
This breaks [multi-part refLists](https://groups.google.com/forum/?fromgroups=#!topic/derbyjs/5xqPYTaGNEA), which I assume you don't use.

**Comment by lefnire** [comment]
Yeah, not using multi-part, so we should be fine. I'll block of some time
hopefully tomorrow to play with your changes,

**Comment by toebu** [comment]
Nice, thanks for all the comments. Good thought to make a github issue for this. I'll only have time on Friday to look at this (exam on Thursday). Let me know @lefnire if you found anything until then.

**Comment by IvanGaravito** [comment]
Hi @lefnire, I been out for a while (offline).

How has been your play with the changes?
Sometime ago I had a similar issue with refLists using `main.templateIds` as indexes and solved including another object level `main.template.ids`. Hope this can help.

By the way, recently had a problem when using queries because some items are not listed and others are repeated. Looking at the database I found that when you remove item(s) it is removed only from the IDs array, so if your query takes only your data collection (e.g. `templates`) then it will show you the ok items and the trashed items.

**Comment by lefnire** [comment]
@SLaks this turned out indeed to be https://github.com/codeparty/derby/issues/215. The solution was static binding, as https://github.com/codeparty/derby/pull/204 didn't seem to fix the problem. Hack-job solution, have to refresh the DOM whenever they add a new tag as a result https://github.com/lefnire/habitrpg/commit/494d5cecf9727b4579be396c0f10855f33f21195 - but I think that this will at least provide some more test-bed for https://github.com/codeparty/derby/issues/215.

Also, tried moving from refLists to arrays - same problem, so you were right

**Comment by toebu** [comment]
Seems like getting rid of the refLists does help something. With the commit above, we can get rid of the reload when adding/removing a tag.

**Comment by lefnire** [comment]
I'm gonna close this ticket, no need for this migration - I know now what the issue is for almost all refLists bugs we've encountered.

Firstly, the DOM issues (tags not associating with correct `#with` / `#each`) is due to https://github.com/codeparty/derby/issues/267. Bug indeed, but that ticket shows a workaround.

The BIG issues we've encountered - duplicates and emptying lists - is because Racer 0.3 was ineffective at knowing when a successful transaction had already gone through, and on a slow connection / server restart it would attempt the transaction again. For push/unshift this would cause duplicates, for `model.remove()` it would keep removing 1 item until the array was emptied, not knowing it had already removed 1 item. This happens for both refLists and arrays. In fact, we're lucky we _used_ refLists (and that we weren't using the option to delete with removal) because when Racer borked the id-list, we could restore it from the `user.tasks` object.

Theoretically this stuff is solved in 0.5's OT algorithms. But I'm still iffy on 0.5 and it's Redis requirements, which is a different conversation. Regardless, this ticket is irrelevant at this point.
